python/tests/test_nn.py

```diff
@@ -792,6 +792,78 @@ def test_log_cosh_loss(self):
         loss = nn.losses.log_cosh_loss(inputs, targets, reduction="mean")
         self.assertAlmostEqual(loss.item(), 0.433781, places=6)
 
+    def test_focal_loss(self):
+        inputs = mx.array([[2.0, -1.0, 3.0, 0.1], [-1.0, 2.0, -0.5, 0.2]])
+        targets = mx.array([[1.0, 0.0, 0.0, 1.0], [0.0, 1.0, 0.0, 1.0]])
```

Comment:
It is better to use original test examples used above. 


Also, it is likely to use pre-defined expected output rather than copy of your loss fuction at losses.py
```python
  ce_loss = mx.logaddexp(0.0, inputs) - targets * inputs
  pt = mx.exp(-ce_loss)
  expected_loss = -alpha * ((1 - pt) ** gamma) * ce_loss
  expected_loss = mx.mean(expected_loss)
  loss = nn.losses.focal_loss(
      inputs, targets, alpha=alpha, gamma=gamma, reduction="mean"
  )
```

These are my codes that uses like `[[0.5, 0.5, 0.2, 0.9], [0.1, 0.3, 0.5, 0.5]]`.  You can extend test functions to tversky_loss and iou. However, I am not sure why less common loss are added on this pr.

Nice work!

cc @awni 

```Python
    def test_dice_loss(self):
        inputs = mx.array([[0.5, 0.5, 0.2, 0.9], [0.1, 0.3, 0.5, 0.5]])
        targets = mx.array([[1, 0, 1, 0], [0, 1, 0, 1]])

        # Test with reduction 'none'
        losses_none = nn.losses.dice_loss(inputs, targets, reduction="none")
        expected_none = mx.array([0.658536, 0.529412])  # Example expected values
        self.assertTrue(mx.allclose(losses_none, expected_none))

        # Test with reduction 'mean'
        losses_mean = nn.losses.dice_loss(inputs, targets, reduction="mean")
        expected_mean = mx.mean(expected_none)
        self.assertTrue(mx.allclose(losses_mean, expected_mean))

        # Test with reduction 'sum'
        losses_sum = nn.losses.dice_loss(inputs, targets, reduction="sum")
        expected_sum = mx.sum(expected_none)
        self.assertTrue(mx.allclose(losses_sum, expected_sum))

    def test_focal_loss(self):
        inputs = mx.array([[0.5, 0.5, 0.2, 0.9], [0.1, 0.3, 0.5, 0.5]])
        targets = mx.array([[1, 0, 1, 0], [0, 1, 0, 1]])
        alpha = 0.25
        gamma = 2.0

        # Test with reduction 'none'
        losses_none = nn.losses.focal_loss(
            inputs, targets, alpha, gamma, reduction="none"
        )
        expected_none = mx.array(
            [
                [0.0433217, 0.0433217, 0.25751, 0.466273],
                [0.000263401, 0.147487, 0.0433217, 0.0433217],
            ]
        )
        self.assertTrue(mx.allclose(losses_none, expected_none))

        # Test with reduction 'mean'
        losses_mean = nn.losses.focal_loss(
            inputs, targets, alpha, gamma, reduction="mean"
        )
        expected_mean = mx.mean(expected_none)
        self.assertTrue(mx.allclose(losses_mean, expected_mean))

        # Test with reduction 'sum'
        losses_sum = nn.losses.focal_loss(
            inputs, targets, alpha, gamma, reduction="sum"
        )
        expected_sum = mx.sum(expected_none)
        self.assertTrue(mx.allclose(losses_sum, expected_sum))

    def test_contrastive_loss(self):
        embeddings1 = mx.array([[0.5, 0.5, 0.2, 0.9], [0.1, 0.3, 0.5, 0.5]])
        embeddings2 = mx.array([[0.6, 0.4, 0.3, 0.8], [0.2, 0.5, 0.6, 0.4]])
        targets = mx.array([1, 0])
        margin = 1.0

        # Test with reduction 'none'
        losses_none = nn.losses.contrastive_loss(
            embeddings1, embeddings2, targets, margin, reduction="none"
        )
        expected_none = mx.array([0.2, 0.735425])
        self.assertTrue(mx.allclose(losses_none, expected_none))

        # Test with reduction 'mean'
        losses_mean = nn.losses.contrastive_loss(
            embeddings1, embeddings2, targets, margin, reduction="mean"
        )
        expected_mean = mx.mean(expected_none)
        self.assertTrue(mx.allclose(losses_mean, expected_mean))

        # Test with reduction 'sum'
        losses_sum = nn.losses.contrastive_loss(
            embeddings1, embeddings2, targets, margin, reduction="sum"
        )
        expected_sum = mx.sum(expected_none)
        self.assertTrue(mx.allclose(losses_sum, expected_sum))
```